src/js/connection.js

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2022, Tony Atkins
+ * Copyright 2011-2020, Colin Clark
+ *
+ *  Licensed under the MIT license, see LICENSE for details.
+ */
+
+(function (fluid) {
+    "use strict";
+    var youme = fluid.registerNamespace("youme");
+
+    /**
+     *
+     * A MIDI Connection represents a connection to a single input or output port. Unless you are completing the wiring
+     * yourself, you should not use this directly, but should use `youme.connection.input` or `youme.connection.output`
+     * instead.
+     *
+     **/
+    fluid.defaults("youme.connection", {
+        gradeNames: ["fluid.modelComponent"],
+
+        model: {
+            port: false,
+            open: false
+        },
+
+        events: {
+            onPortOpen: null,
+            onPortClose: null,
+            onError: null
+        },
+
+        listeners: {
+            "onError.logError": {
+                funcName: "fluid.log",
+                args: [fluid.logLevel.WARN, "{arguments}.0"]
+            }
+        },
+
+        modelListeners: {
+            open: {
+                funcName: "youme.connection.updateConnectionState",
+                args: ["{that}.model.port", "{that}.model.open", "{that}.events.onPortOpen.fire", "{that}.events.onPortClose.fire", "{that}.events.onError.fire"] // port, isNowOpen, onPortOpen, onPortClose, onError
+            },
+            port: {
+                excludeSource: "init", // On startup, both open and port are set, so we let the open listener decide whether to open the port.
+                funcName: "youme.connection.handlePortChange",
+                args: ["{change}.oldValue", "{that}.model.port", "{that}.model.open", "{that}.events.onPortOpen.fire", "{that}.events.onPortClose.fire", "{that}.events.onError.fire"] // oldPort, newPort, isOpen, onPortOpen, onPortClose, onError
+            }
+        }
+    });
+
+    youme.connection.handlePortChange = function (oldPort, newPort, isOpen, onPortOpen, onPortClose, onError) {
+        if (oldPort) {
+            oldPort.close();
```

Comment:
Looks like there is a risk of a race here since the port closing operation is asynchronous and the promise return value is lost here. It seems best to turn handlePortChange into a transforming promise chain which closes the old port in one listener and opens the new one in another. Failing that, this can become a regular async function.

Reply:
I ended up having to use some of the old conventions to ensure that rejection would fire `onError`, but did at least line these up as links in a chain and ensure that they return promises.